test/soak/main_test.go

```diff
@@ -0,0 +1,233 @@
+		// Enables KnativeService in the scenario.
+		//eventshub.WithKnativeServiceForwarder,
+		environment.WithPollTimings(Flags.PollInterval, Flags.PollDuration),
+		environment.WithTestLogger(t),
+	)
+}
+
+type SoakFn func(context.Context, environment.Environment, *testing.T)
+
+type SoakTest struct {
+	/*
+		Prefix to be used for namespaces. Actual test namespace is <prefix><copyId>
+	*/
+	NamespacePrefix string
+	/*
+		Function invoked during setup for each soak test copy.
+		Use env.Test(ctx, t, f)  inside to handle cleanup of the resources created by the features.
+	*/
+	SetupFn SoakFn
+	/*
+		Function invoked for each iteration.
+		Use env.Test(ctx, t, featureFn(SoakEnvFromContext(ctx)))  inside to dynamically create feature based on SoakEnv copy and iteration.
+		Soak env does handle cleanup of all resources created by the features during the iteration
+	*/
+	IterationFn SoakFn
+	/*
+		Function invoked after the iterations complete. Could be used to teardown or state verification after the end of the soak test.
+	*/
+	TeardownFn SoakFn
+}
+
+type soakKey struct{}
+
+type SoakEnv interface {
+	CopyID() int
+	Iteration() int
+	Namespace() string
+}
+
+type soakEnvImpl struct {
+	copyID    int
+	iteration int
+	namespace string
+}
+
+func (env *soakEnvImpl) CopyID() int {
+	return env.copyID
+}
+
+func (env *soakEnvImpl) Iteration() int {
+	return env.iteration
+}
+
+func (env *soakEnvImpl) Namespace() string {
+	return env.namespace
+}
+
+func soakEnvImplFromContext(ctx context.Context) *soakEnvImpl {
+	if se, ok := ctx.Value(soakKey{}).(*soakEnvImpl); ok {
+		return se
+	}
+	panic("no soak environment found in the context, make sure you're executing a soak test code within a RunSoakTest")
+}
+
+func SoakEnvFromContext(ctx context.Context) SoakEnv {
+	return soakEnvImplFromContext(ctx)
+}
+
+func RunSoakTest(t *testing.T, test SoakTest, copies int) {
+	for copyID := 0; copyID < copies; copyID++ {
+		copyID := copyID
+		namespace := test.NamespacePrefix + strconv.Itoa(copyID)
+		t.Run(namespace, func(t *testing.T) {
+			t.Parallel()
+
+			since := time.Now()
+
+			ctx, env := soakTestEnvironment(t, namespace)
+
+			// Execute the setup "features", store the references created during setup for cleanup at the end
+			setupCtx := context.WithValue(ctx, soakKey{}, &soakEnvImpl{
+				copyID:    copyID,
+				iteration: -1,
+				namespace: namespace,
+			})
+
+			if test.SetupFn != nil {
+				test.SetupFn(setupCtx, env, t)
+			}
+			if t.Failed() {
+				feature.LogReferences(env.References()...)(ctx, t)
+				return
+			}
+
+			iteration := 0
+			// Repeat the soak test for the duration specified by --soak-duration flag
+			for since.Add(Flags.Duration).After(time.Now()) {
+				// During each iteration, generate the "iteration" features and run them as Tests
+				// Cleanup all resources for these features at the end of the iteration
+
+				// TODO: can we do this? (use an empty env for the iteration, but keep using the original context?)
```

Comment:
Do you want to keep the comment here? We has some discussion around that.

Reply:
well, if we agree we _can_ do it like this... I don't see any alternative besides adding this as an API to reconciler framework

Reply:
I think the tests showed it can be done like this. But if we can formalize the requirements and it would be useful for REKT framework in general, we can contribute this upstream later, and possibly re-use it here. But for now I'd just just what we have here.

Reply:
Maybe we can create an issue upstream in https://github.com/knative-extensions/reconciler-test if the requirements are clear.